src/library/scala/collection/SeqLike.scala

```diff
@@ -504,15 +504,19 @@ trait SeqLike[+A, +Repr] extends Any with IterableLike[A, Repr] with GenSeqLike[
    *  @return  A new $coll which contains the first occurrence of every element of this $coll.
    */
   def distinct: Repr = {
-    val b = newBuilder
-    val seen = mutable.HashSet[A]()
-    for (x <- this) {
-      if (!seen(x)) {
-        b += x
-        seen += x
+    val size = this.sizeHintIfCheap
+    if (size == 0 || size == 1 || isEmpty) repr
```

Comment:
Does it make sense to special-case this stuff rather than using `lengthCompare` (which should be just as fast for `List` and doesn't rely on overriding `sizeHintIfCheap` in other LinearSeqs)? The only semantic difference I can think of is that `lengthCompare` would force a lazy collection but you have to do that anyway.

Reply:
Well, it will also create one more object since `SeqLike.lengthCompare` uses iterator.
If it matters or not is separate question...

Reply:
List.lengthCompare use LinearSeqOptimised, so no iterator

Reply:
lengthCompare would also allow for the trivial case of length 2 (which seems to be 5% of the cases in the test data
e.g. 
```
val compared = lengthCompare(2)
if (compared < 0) repr
else if (compared == 0) {
  val it = iterator
  val v1 = it.next
  val v2 = it.next
  if (v1 != v2)  repr else {
     val b = newBuilder 
     b += v1
     b.build // no faster way to build a single element collection?
  }
} else {
... existing code
```

Reply:
@szeiger Do you think it would be valuable to add this code to handle special case of two elements?
If not then I do not see big advantage in rewriting this to `lengthCompare` but i can do this if it's your preference.
In other words: what do you think we need to merge this?

Reply:
I discussed with @szeiger and agreed that we would move to lengthCompare based